### ISSUE TITLE
href path not recognize children path from route object declare outside typesafeBrowserRouter

### ISSUE THREAD

**Issue by aviram5** [stargaze-co/react-router-typesafe#14]
this is my typesafeBrowserRouter:
```ts
export const { router, href } = typesafeBrowserRouter([
  {
    path: "/",
    element: <AppLoading />,
  },
  {
    path: "app",
    children: AppRoutes,
  },
]);
```

AppRoutes is a simple RouteObject[] as follow: 
```ts
const AppRoutes: RouteObject[] = [
  {
    index: true,
    async lazy() {
      const { Home } = await import("../pages");
      return { Component: Home };
    },
  },
  {
    path: "contact",
    async lazy() {
      const { Contact } = await import("../pages");
      return {
        element: <ProtectedRoute children={<Contact />} />,
      };
    },
  },
];
```
when using href it doesn't recognize "/app/contact" as a route it's only options are "/"|"app"
`<NavLink to={href({ path: "app/contact", })}>contact page</NavLink>`
error: 

> Type '"app/contact"' is not assignable to type '"/" | "app"'

I would love to know if there is a way to fix this other than setting everything inside the typesafeBrowserRouter

**Comment by fredericoo** [comment]
When you annotate it as `RouteObject[]`, you’re telling typescript to expect the widest declaration possible of that type!

If you were to instead use the `satisfies` keyword, you’d get the same type safety but it’d be narrowing your paths correctly:
```ts
const AppRoutes = [
  {
    index: true,
    async lazy() {
      const { Home } = await import("../pages");
      return { Component: Home };
    },
  },
  {
    path: "contact",
    async lazy() {
      const { Contact } = await import("../pages");
      return {
        element: <ProtectedRoute children={<Contact />} />,
      };
    },
  },
] as const satisfies RouteObject[];
```

give this a try please and tell me how it went!

**Comment by aviram5** [comment]
hey sorry for the delay.
i tried your suggestion but sadly it still won't work.
i even tried to export your library NarrowArray and NarrowKey types and used them but no luck.
i ditched this approach even though it's more readable but without the type safety it's not worth it.
if you have other suggestions i would love to try them

**Comment by fredericoo** [comment]
Hey! So i had forgotten to tell you to add as const as well, leaving it `as const satisfies RouteObject[]`.

There are one I need to reiterate and *are about react router*: always specify the full path in routes: children routes do not necessarily nest paths.

I made a codesandbox for you to see a working example:
https://codesandbox.io/p/sandbox/typesafe-routes-2mnqqw?file=%2Fsrc%2Findex.tsx

You did however bring to light an issue with the `href` function: if the routes are declared without prefixing them with `/`, the output href will also be missing the slash, making it so that routes are relative to their current ones. This is a bug and I’ll be issuing a fix soon. For the time being, **prefix your routes with `/`** to get them working as expected. I’d recommend doing that anyway, to be fair

**Comment by aviram5** [comment]
so i tried with "as const":
```
const AppRoutes = [
  {
    index: true,
    async lazy() {
      const { Home } = await import("../pages");
      return { Component: Home };
    },
  },
  {
    path: "/app/contact",
    async lazy() {
      const { Contact } = await import("../pages");
      return {
        element: <ProtectedRoute children={<Contact />} />,
      };
    },
  },
] as const satisfies RouteObject[];

export const { router, href } = typesafeBrowserRouter([
  {
    path: "/",
    children: [
      {
        index: true,
        element: <AppLoading />,
      },
      {
        path: "/app",
        element: <MainLayout />,
        children: AppRoutes,
      },
    ],
  },
]);
```

but still when using href i have no suggestions:

![צילום מסך 2024-04-02 140354](https://github.com/fredericoo/react-router-typesafe/assets/60235058/05440b3a-f1a7-4866-b151-57ea00eb13d8)

when everything is defined inside the typesafeBrowserRouter: 
```
export const { router, href } = typesafeBrowserRouter([
  {
    path: "/",
    children: [
      {
        index: true,
        element: <AppLoading />,
      },
      {
        path: "/app",
        element: <MainLayout />,
        children: [
          {
            index: true,
            async lazy() {
              const { Home } = await import("../pages");
              return { Component: Home };
            },
          },
          {
            path: "/app/contact",
            async lazy() {
              const { Contact } = await import("../pages");
              return {
                element: <ProtectedRoute children={<Contact />} />,
              };
            },
          },
        ],
      },
    ],
  },
]);
```
![צילום מסך 2024-04-02 142302](https://github.com/fredericoo/react-router-typesafe/assets/60235058/1a0e8c07-5659-41d3-8349-ba11aac88407)

**Comment by fredericoo** [comment]
Thanks for the detailed report! I found the issue: the route layer that had the children set to `AppRoutes` was widening its type to `string`. I can probably sprinkle a const generic there to make it work.
![CleanShot 2024-04-02 at 12 52 10@2x](https://github.com/fredericoo/react-router-typesafe/assets/45042736/e8b0c9f3-1bd3-4d1b-8d04-a73e7dd0a339)

If you want to make it work before this fix is issued, you can also slap an `as const` like so:
```ts
{
-        path: "/app",
+        path: "/app" as const,
        element: <MainLayout />,
        children: AppRoutes,
      },
```

I’ll be working on both fixes tonight or tomorrow ◡̈ tysm

**Comment by aviram5** [comment]
adding the "as const" to the path property work as expected.
thank you :) really appreciate your work!

**Comment by fredericoo** [comment]
The two issues should be fixed in v1.5.0! Made sure to add tests to prevent this from regressing, as it’s very very subtle ◡̈ 

Thanks a lot for being so attentive and supportive in giving examples. This is how we get it fixed with ease.
